controllers/topolvm_node.go

```diff
@@ -0,0 +1,321 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	lvmv1alpha1 "github.com/red-hat-storage/lvm-operator/api/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"path/filepath"
+	cutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"strings"
+)
+
+const (
+	topolvmNodeName = "topolvm-node"
+)
+
+type topolvmNode struct{}
+
+func (n topolvmNode) getName() string {
+	return topolvmNodeName
+}
+
+func (n topolvmNode) ensureCreated(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	nodeDaemonSet := getNodeDaemonSet(lvmCluster)
+	result, err := cutil.CreateOrUpdate(ctx, r.Client, nodeDaemonSet, func() error {
+		// make sure LVMCluster CR garbage collects this daemonset and also block owner removal
+		return cutil.SetControllerReference(lvmCluster, nodeDaemonSet, r.Scheme)
+	})
+
+	switch result {
+	case cutil.OperationResultCreated:
+		r.Log.Info(topolvmNodeName, "operation", result, "name", nodeDaemonSet.Name)
+	case cutil.OperationResultUpdated:
+		r.Log.Info(topolvmNodeName, "operation", result, "name", nodeDaemonSet.Name)
+	case cutil.OperationResultNone:
+		r.Log.Info(topolvmNodeName, "operation", result, "name", nodeDaemonSet.Name)
+	default:
+		r.Log.Error(err, fmt.Sprintf("%s reconcile failure", topolvmNodeName), "name", nodeDaemonSet.Name)
+		return err
+	}
+
+	return nil
+}
+
+// ensureDeleted should wait for the resources to be cleaned up
+func (n topolvmNode) ensureDeleted(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	NodeDaemonSet := &appsv1.DaemonSet{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: TopolvmNodeDaemonsetName, Namespace: lvmCluster.Namespace}, NodeDaemonSet)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("topolvm node deleted", "TopolvmNode", NodeDaemonSet.Name)
+			return nil
+		}
+		r.Log.Error(err, "unable to retrieve topolvm node daemonset", "TopolvmNode", NodeDaemonSet.Name)
+		return err
+	}
+	return nil
+}
+
+// updateStatus should optionally update the CR's status about the health of the managed resource
+// each unit will have updateStatus called individually so
+// avoid status fields like lastHeartbeatTime and have a
+// status that changes only when the operands change.
+func (n topolvmNode) updateStatus(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	return nil
+}
+
+func getNodeDaemonSet(lvmCluster *lvmv1alpha1.LVMCluster) *v1.DaemonSet {
+	hostPathDirectory := corev1.HostPathDirectory
+	hostPathDirectoryOrCreateType := corev1.HostPathDirectoryOrCreate
+	storageMedium := corev1.StorageMediumMemory
+
```

Comment:
Looks like a lot of static code when generating the daemonset spec. I've a suggestion to use to yaml files and read them directly using go embed? This will reduce the lines of codes and also make it easier to understand the spec.  An example of this approach can be seen [here](https://github.com/openshift/local-storage-operator/tree/master/assets)

Reply:
I would prefer to keep this in the code for now. We can revisit this later.